src/network/networkd-dhcp4.c

```diff
@@ -1449,6 +1461,96 @@ int config_parse_dhcp_request_options(
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+
+        Network *network = data;
+        const char *x;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        if (isempty(rvalue)) {
+                network->dhcp_send_options = ordered_hashmap_free_with_destructor(network->dhcp_send_options, sd_dhcp_option_unref);
+                return 0;
+        }
+
+        r = ordered_hashmap_ensure_allocated(&network->dhcp_send_options, &trivial_hash_ops);
+        if (r < 0)
+                return log_oom();
+
+        for (x = rvalue;;) {
+                _cleanup_free_ char *word = NULL;
+                sd_dhcp_option *p;
+                uint32_t u;
+                char *split;
+
+                r = extract_first_word(&x, &word, NULL, EXTRACT_UNQUOTE);
+                if (r == 0)
+                        return 0;
+                if (r == -ENOMEM)
+                        return log_oom();
+                if (r < 0) {
+                        log_syntax(unit, LOG_ERR, filename, line, r, "Failed to extract first word, ignoring: %s", rvalue);
+                        return 0;
+                }
+
+                split = strchr(word, ':');
+                if (!split)
+                        continue;
+
+                *split = '\0';
+
+                r = safe_atou32(word, &u);
+                if (r < 0) {
+                        log_syntax(unit, LOG_ERR, filename, line, r,
+                                   "DHCP send option is invalid, ignoring assignment: %s", rvalue);
+                        return 0;
+                }
+
+                if (u < 1 || u >= 255) {
+                        log_syntax(unit, LOG_ERR, filename, line, r,
+                                   "DHCP send option is invalid, valid range is 1-254, ignoring assignment: %s", rvalue);
+                        return 0;
+                }
+
+                if (sd_dhcp_option_new(&p) < 0)
+                        return log_oom();
+
+                r = sd_dhcp_set_dhcp_option(p, u, split + 1, strlen(split + 1));
+                if (r < 0) {
+                        log_syntax(unit, LOG_ERR, filename, line, r,
+                                   "Failed to store DHCP request option '%s', ignoring assignment: %m", rvalue);
+                        return 0;
+                }
+
+                r = ordered_hashmap_put(network->dhcp_send_options, UINT_TO_PTR(u), p);
+                if (r == -EEXIST) {
+                        log_syntax(unit, LOG_ERR, filename, line, r,
+                                   "Failed to store DHCP request option '%s' already exist.", word);
+                        continue;
```

Comment:
OK, let's talk about semantics here. You use the option number a key in the hashmap, and the first value that is set wins. You said that it is OK to have the same option appear multiple times, as long as the rhs is different. This needs to be consistent. If `sd_dhcp_client_add_dhcp_option()` allows multiple options with the same value, then this function must too.

In other words, it is pointless to use a Hashmap here and then push values into a Set.

Reply:
This need values needs to be appended what I think